src/main/drivers/system_stm32g4xx.c

```diff
@@ -0,0 +1,207 @@
+void forcedSystemResetWithoutDisablingCaches(void)
+{
+    persistentObjectWrite(PERSISTENT_OBJECT_RESET_REASON, RESET_FORCED);
+
+    __disable_irq();
+    NVIC_SystemReset();
+}
+
+void systemResetToBootloader(bootloaderRequestType_e requestType)
+{
+    switch (requestType) {
+#if defined(USE_FLASH_BOOT_LOADER)
+    case BOATLOADER_REQUEST_FLASH:
+        persistentObjectWrite(PERSISTENT_OBJECT_RESET_REASON, RESET_FLASH_BOOTLOADER_REQUEST);
+
+        break;
+#endif
+    case BOOTLOADER_REQUEST_ROM:
+    default:
+        persistentObjectWrite(PERSISTENT_OBJECT_RESET_REASON, RESET_BOOTLOADER_REQUEST_ROM);
+
+        break;
+    }
+
+    __disable_irq();
+    NVIC_SystemReset();
+}
+
+#define SYSMEMBOOT_VECTOR_TABLE ((uint32_t *)0x1fff0000)
+
+typedef void *(*bootJumpPtr)(void);
+
+void systemJumpToBootloader(void)
+{   
+    __SYSCFG_CLK_ENABLE();
+    
+    uint32_t bootStack =  SYSMEMBOOT_VECTOR_TABLE[0];
+    
+    bootJumpPtr SysMemBootJump = (bootJumpPtr)SYSMEMBOOT_VECTOR_TABLE[1];
+    
+    __set_MSP(bootStack); //Set the main stack pointer to its default values
+    
+    SysMemBootJump();
+    
+    while (1);
+}
+
+static uint32_t bootloaderRequest;
+
+void systemCheckResetReason(void)
+{
+    bootloaderRequest = persistentObjectRead(PERSISTENT_OBJECT_RESET_REASON);
+
+    switch (bootloaderRequest) {
+#if defined(USE_FLASH_BOOT_LOADER)
+    case BOATLOADER_REQUEST_FLASH:
+#endif
+    case RESET_BOOTLOADER_REQUEST_ROM:
+        persistentObjectWrite(PERSISTENT_OBJECT_RESET_REASON, RESET_BOOTLOADER_POST);
+        break;
+
+    case RESET_MSC_REQUEST:
+        // RESET_REASON will be reset by MSC
+        return;
+
+    case RESET_FORCED:
+        persistentObjectWrite(PERSISTENT_OBJECT_RESET_REASON, RESET_NONE);
+        return;
+
+    case RESET_NONE:
+        if (!(RCC->CSR & RCC_CSR_SFTRSTF)) {
+            // Direct hard reset case
+            return;
+        }
+        // Soft reset; boot loader may have been active with BOOT pin pulled high.
+        FALLTHROUGH;
+
+    case RESET_BOOTLOADER_POST:
+        // Boot loader activity magically prevents SysTick from interrupting.
+        // Issue a soft reset to prevent the condition.
+        forcedSystemResetWithoutDisablingCaches(); // observed that disabling dcache after cold boot with BOOT pin high causes segfault.
+    }
+
+    void (*SysMemBootJump)(void);
+    __SYSCFG_CLK_ENABLE();
+
+    systemJumpToBootloader();
+
+#define SYSTEM_BOOTLOADER_VEC 0x1fff0000
+
+    uint32_t p =  (*((uint32_t *)SYSTEM_BOOTLOADER_VEC));
+    __set_MSP(p); //Set the main stack pointer to its defualt values
+    SysMemBootJump = (void (*)(void)) (*((uint32_t *)(SYSTEM_BOOTLOADER_VEC + 4))); // Point the PC to the System Memory reset vector (+4)
+    SysMemBootJump();
+    while (1);
+}
+
+// Nucleo-G474RE board seems to come with software BOOT0 enabled.
+// Call this function once from init() to honor PB8-BOOT0 pin status for boot loader invocation.
+void systemBOOT0PinBootLoaderEnable(void)
```

Comment:
Where is this used?

Reply:
Oh, this is a code that should be run only once for a brand new out of a package Nucleo-G474RE.

Reply:
May be keep the code inside a special conditional, say, `RUN_ONCE_FROM_INIT_FOR_NUCLEO_G474RE_BOARD`?

Thoughts?